server/src/main/java/org/elasticsearch/index/mapper/RuntimeField.java

```diff
@@ -175,4 +177,31 @@ RuntimeField parse(String name, Map<String, Object> node, Mapper.TypeParser.Pars
         }
         return Collections.unmodifiableMap(runtimeFields);
     }
+
+    /**
+     * Collect and return all {@link MappedFieldType} exposed by the provided {@link RuntimeField}s.
+     * Note that validation is performed to make sure that there are no name clashes among the collected runtime fields.
+     * This is because runtime fields with the same name are not accepted as part of the same section.
+     * @param runtimeFields the runtime to extract the mapped field types from
+     * @return the collected mapped field types
+     */
+    static Map<String, MappedFieldType> collectFieldTypes(Collection<RuntimeField> runtimeFields) {
+        return runtimeFields.stream()
+            .flatMap(runtimeField -> {
+                List<String> names = runtimeField.asMappedFieldTypes().stream().map(MappedFieldType::name)
+                    .filter(name -> name.equals(runtimeField.name()) == false
+                        && (name.startsWith(runtimeField.name() + ".") == false
+                        || name.length() > runtimeField.name().length() + 1 == false))
+                    .collect(Collectors.toList());
+                if (names.isEmpty() == false) {
+                    throw new IllegalStateException("Found sub-fields with name not belonging to the parent field they are part of "
+                        + names);
+                }
+                return runtimeField.asMappedFieldTypes().stream();
+            })
+            .collect(Collectors.toUnmodifiableMap(MappedFieldType::name, mappedFieldType -> mappedFieldType,
+                (t, t2) -> {
+                    throw new IllegalArgumentException("Found two runtime fields with same name [" + t.name() + "]");
```

Comment:
Was this enforced at all before, or would one runtime field just override another? If not it could be worth calling out in the 'migration' docs in case some user runs into it.

Reply:
Good question, I also wondered for a bit. This scenario could not happen before, because there was only one way to define a field with a specific name (assuming there are no duplicated keys in the json request in which case I think the result is not  deterministic). It was only technically possible if `asMappedFieldType` returned a name that differs from the `RuntimeField#name`, but this was never the case until now at least with our implementations. I am leaning towards nothing to do here, this could really only happen with a poorly implemented custom runtime field (I haven't tested it but I assume that would be possible)?

Reply:
Indeed it seems like it'd be very rare, because in most cases we'll fail to parse the JSON. Probably not worth calling out. 

In case it's interesting context, it's actually possible to define the same field twice through mixed object notation. With 'standard' mappings, several users have run into this and it caused confusion to the point where we proposed disallowing it (https://github.com/elastic/elasticsearch/issues/37437).

Reply:
I am aware that this is possible in the properties section, but it is not in the runtime section as it is rather flat and objects are not supported there.

Reply:
Ah thanks for clarifying, I hadn't realized objects weren't even supported.